test/e2e/federated-secret.go

```diff
@@ -18,69 +18,161 @@ package e2e
+	var (
+		clusters       map[string]*cluster // All clusters, keyed by cluster name
+		federationName string
+	)
+
 	f := framework.NewDefaultFederatedFramework("federated-secret")
 
 	Describe("Secret objects", func() {
-		AfterEach(func() {
-			framework.SkipUnlessFederated(f.Client)
 
-			nsName := f.FederationNamespace.Name
-			// Delete registered secrets.
-			// This is if a test failed, it should not affect other tests.
-			secretList, err := f.FederationClientset_1_4.Core().Secrets(nsName).List(api.ListOptions{})
-			Expect(err).NotTo(HaveOccurred())
-			for _, secret := range secretList.Items {
-				err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
-				Expect(err).NotTo(HaveOccurred())
+		BeforeEach(func() {
+			framework.SkipUnlessFederated(f.Client)
+			if federationName = os.Getenv("FEDERATION_NAME"); federationName == "" {
+				federationName = DefaultFederationName
 			}
+			clusters = map[string]*cluster{}
+			_ = registerClusters(clusters, UserAgentName, federationName, f)
+		})
+
+		AfterEach(func() {
+			unregisterClusters(clusters, f)
 		})
 
 		It("should be created and deleted successfully", func() {
 			framework.SkipUnlessFederated(f.Client)
-
 			nsName := f.FederationNamespace.Name
-			secret := createSecretOrFail(f.FederationClientset_1_4, nsName)
-			By(fmt.Sprintf("Creation of secret %q in namespace %q succeeded.  Deleting secret.", secret.Name, nsName))
-			// Cleanup
-			err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
-			framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, secret.Namespace)
-			By(fmt.Sprintf("Deletion of secret %q in namespace %q succeeded.", secret.Name, nsName))
-		})
+			secret, _ := createSecretOrFail(f.FederationClientset_1_4, nsName)
 
+			defer func() { // Cleanup
+				By(fmt.Sprintf("Deleting secret %q in namespace %q", secret.Name, nsName))
+				err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
+				framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, nsName)
+			}()
+			// wait for secret shards being created
+			waitForSecretShardsOrFail(nsName, secret, clusters)
+			secret, _ = updateSecretOrFail(f.FederationClientset_1_4, nsName)
+			waitForSecretShardsUpdatedOrFail(nsName, secret, clusters)
+		})
 	})
 })
 
-func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1.Secret {
-	if clientset == nil || len(namespace) == 0 {
-		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteSecretOrFail: clientset: %v, namespace: %v", clientset, namespace))
-	}
-	By(fmt.Sprintf("Creating federated secret %q in namespace %q", FederatedSecretName, namespace))
-
+func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) (*v1.Secret, error) {
```

Comment:
You seem to be ignoring the returned err everywhere assuming the function will fail if there is an error.
No need to return an error in that case?

Same for the updateSecret fn